membership-attribute-service/app/services/zuora/payment/ZuoraPaymentService.scala

```diff
@@ -35,47 +36,22 @@ class PaymentService(zuoraService: ZuoraSoapService, planMap: Map[ProductRatePla
   )(sub: Subscription[SubscriptionPlan.Paid])(implicit logPrefix: LogPrefix): Future[PaymentDetails] = {
     val currency = sub.plan.charges.currencies.head
     // I am not convinced this function is very safe, hence the option
-    val lastPaymentDate = zuoraService
+    val eventualMaybeLastPaymentDate = zuoraService
       .getPaymentSummary(sub.name, currency)
       .map(_.current.serviceStartDate.some)
       .recover { case _ => None }
-    lastPaymentDate.onComplete {
-      case Failure(exception) =>
-        val message = scrub"Failed to get last payment date for sub $sub"
-        logger.error(message, exception)
-      case Success(_) => logger.info(s"Successfully got payment details for $sub")
-    }
+    eventualMaybeLastPaymentDate.withLogging(s"lastPaymentDate for $sub")
 
-    val nextPaymentDate = sub.plan match {
-      case _: Contributor =>
-        sub.plan.chargedThrough.getOrElse(
-          sub.plan.start,
-        ) // If the user has updated their contribution amount via MMA, there may be no charged through date
-      case _ => sub.plan.chargedThrough.getOrElse(sub.acceptanceDate) // If the user hasn't paid us yet, there's no charged through date
-    }
-    val futureMaybePrice = sub.plan.product.name match {
-      // There's no need to get the billing schedule for membership - since there are no promos etc.
-      case "membership" => {
-        Future.successful(sub.plan.charges.price.prices.headOption)
-      }
-      case _ => {
-        val schedule = billingSchedule(sub.id, sub.accountId, 15).map { maybeSchedule =>
-          maybeSchedule.map(schedule => Price(schedule.first.amount, currency))
-        }
-        schedule.onComplete {
-          case Failure(exception) =>
-            val message = scrub"Failed to get billing schedule for sub $sub"
-            logger.error(message, exception)
-          case Success(_) => logger.info(s"Successfully got billing schedule for $sub")
-        }
-        schedule
-      }
-    }
-    (futureMaybePrice |@| getPaymentMethod(sub.accountId, defaultMandateIdIfApplicable) |@| lastPaymentDate) {
-      case (maybePrice, paymentMethod, lpd) => {
-        val nextPayment = maybePrice.map { price => Payment(price, nextPaymentDate) }
-        PaymentDetails(sub, paymentMethod, nextPayment, lpd)
-      }
+    for {
+      account <- zuoraService.getAccount(sub.accountId)
+      eventualMaybeBill = getNextBill(sub.id, account, 15).withLogging(s"next bill for $sub")
```

Comment:
we used to avoid calling `billingSchedule` (now called `getNextBill`) where not necessary, e.g. membership and recurring contribution, this is because they behave(d) predictably, so we could optimise away the call to zuora.
However now contribution can be discounted (suspended) and membership is not such a mainstream product, it makes sense to simplify the code at the cost of the extra zuora call.